### PR TITLE
Introduce brave://adblock-internals

### PR DESCRIPTION
**Pull request by atuchin-m**
Resolves https://github.com/brave/brave-browser/issues/27966

A debug page that shows all active regexs in adblock-rust and the connected memory metrics.
* The page should make the long term adblock memory consumption easy to maintains/understood.
* After https://github.com/brave/adblock-rust/pull/240 we also have a mechanism of **auto-discarding** a regex that was unused for some time. The page helps to visualize that mechanism.
* It also supports discarding a particular regex to see the memory difference. That way we have the ability to estimate memory usage by a particular regex(or a group) and find the most expensive ones.
* Also, the PR adds an ability to override the default discarding policy by a new `base::Feature`. This includes completely disable the feature if needed.

image

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/adblock_rust_ffi/Cargo.toml [hunk 1]
```diff
@@ -5,7 +5,8 @@ authors = ["Brian R. Bondy <netzen@gmail.com>"]
 edition = "2018"
 
 [dependencies]
-adblock = { version = "0.6.3", default-features = false, features = ["full-regex-handling", "object-pooling", "unsync-regex-caching"] }
+# TODO: Revert
```
Comment (atuchin-m) [inline components/adblock_rust_ffi/Cargo.toml:8]: Will be reverted after https://github.com/brave/brave-browser/issues/27967 (before merging this PR)

**Review by mkarolin — APPROVED** [review]
`chromium_src` LGTM

File: components/adblock_rust_ffi/src/lib.rs [hunk 2]
```diff
@@ -343,6 +344,83 @@ pub unsafe extern "C" fn filter_list_metadata_destroy(metadata: *mut FilterListM
...
+) {
+    assert!(!debug_info.is_null());
+    let info = Box::leak(Box::from_raw(debug_info));
+    let regex_data = &info.blocker_debug_info.regex_data;
+    assert!(index < regex_data.len());
+    let entry = &regex_data[index];
+
+    *id = entry.id;
+    *regex = CString::new(
+        entry.regex.as_ref().unwrap_or(&"".to_string()).as_str())
```
Comment (antonok-edm) [inline components/adblock_rust_ffi/src/lib.rs:387]:
```rust
CString::new(entry.regex.as_deref().unwrap_or(""))
```

**Review by antonok-edm — APPROVED** [review]

File: components/adblock_rust_ffi/Cargo.toml (hunk 1 above)
Comment (antonok-edm) [reply components/adblock_rust_ffi/Cargo.toml:8]: we could also include the version bump as part of this PR, if that's easier

File: components/adblock_rust_ffi/src/lib.rs [hunk 3]
```diff
@@ -3,9 +3,10 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this file,
  * You can obtain one at https://mozilla.org/MPL/2.0/. */
 
-use adblock::engine::Engine;
+use adblock::engine::{Engine, EngineDebugInfo};
 use adblock::lists::FilterListMetadata;
 use adblock::resources::{MimeType, Resource, ResourceType};
+use adblock::filters::regex_manager::RegexManagerDiscardPolicy;
```
Comment (antonok-edm) [inline components/adblock_rust_ffi/src/lib.rs:9]: Heads up about https://github.com/brave/adblock-rust/commit/031882d3ca0b4f31d9404a5b8d2af924705aaee4 - this will need to be `adblock::regex_manager::RegexManagerDiscardPolicy`

File: components/adblock_rust_ffi/Cargo.toml (hunk 1 above)
Comment (antonok-edm) [reply components/adblock_rust_ffi/Cargo.toml:8]: (published [`0.7.0`](https://crates.io/crates/adblock/0.7.0))

Comment (atuchin-m) [reply components/adblock_rust_ffi/Cargo.toml:8]: Thanks, updated.

File: components/adblock_rust_ffi/src/lib.rs (hunk 2 above)
Comment (atuchin-m) [reply components/adblock_rust_ffi/src/lib.rs:387]: Done.

File: components/brave_shields/browser/ad_block_engine.cc [hunk 4]
```diff
@@ -178,6 +178,41 @@ bool AdBlockEngine::TagExists(const std::string& tag) {
   return base::Contains(tags_, tag);
 }
 
+base::Value AdBlockEngine::GetDebugInfo() {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+  const auto debug_info_struct = ad_block_client_->getAdblockDebugInfo();
+  base::Value regex_list(base::Value::Type::LIST);
+  for (const auto& regex_entry : debug_info_struct.regex_data) {
+    base::Value regex_info(base::Value::Type::DICTIONARY);
+    regex_info.SetKey("id", base::Value(std::to_string(regex_entry.id)));
```
Comment (boocmp) [inline components/brave_shields/browser/ad_block_engine.cc:187]: SetKey // DEPRECATED: Prefer `Value::Dict::Set()`.

File: components/brave_shields/browser/ad_block_service.cc [hunk 5]
```diff
@@ -313,6 +323,42 @@ void AdBlockService::EnableTag(const std::string& tag, bool enabled) {
...
+  GetTaskRunner()->PostTaskAndReplyWithResult(
+      FROM_HERE,
+      base::BindOnce(&AdBlockEngine::GetDebugInfo,
+                     base::Unretained(default_engine_.get())),
+      base::BindOnce(&AdBlockService::OnGetDebugInfoFromDefaultEngine,
+                     weak_factory_.GetWeakPtr(), std::move(callback)));
+}
+
+void AdBlockService::DiscardRegex(uint64_t regex_id) {
+  // Dispath to both default & additional engines, ids are unique.
```
Comment (boocmp) [inline components/brave_shields/browser/ad_block_service.cc:341]: dispatCh ?

**Review by boocmp — APPROVED** [review]

File: components/adblock_rust_ffi/src/lib.rs (hunk 3 above)
Comment (atuchin-m) [reply components/adblock_rust_ffi/src/lib.rs]: Done

File: components/brave_shields/browser/ad_block_engine.cc (hunk 4 above)
Comment (atuchin-m) [reply components/brave_shields/browser/ad_block_engine.cc]: Thanks, fixed

File: components/brave_shields/browser/ad_block_service.cc (hunk 5 above)
Comment (atuchin-m) [reply components/brave_shields/browser/ad_block_service.cc]: Fixed.

**Review by iefremov — APPROVED** [review]

File: build/rust/Cargo.lock [hunk 6]
```diff
@@ -4,9 +4,9 @@ version = 3
 
 [[package]]
 name = "adblock"
-version = "0.6.3"
+version = "0.7.0"
```
Comment (goodov) [inline build/rust/Cargo.lock:7]: is this change intentional?

File: browser/ui/webui/brave_adblock_internals_ui.cc [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+namespace {
+
+// See chrome/browser/metrics/process_memory_metrics_emitter.cc.
+struct MemoryMetric {
+  // The root dump name that represents the required metric.
+  const char* const dump_name;
+  // The type of metric that is measured, usually size in bytes or object count.
+  const char* const metric;
+} kCollectedMemoryMetrics[] = {
```
Comment (goodov) [inline browser/ui/webui/brave_adblock_internals_ui.cc:35]: `const`

File: browser/ui/webui/brave_adblock_internals_ui.cc [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+#include "brave/browser/brave_browser_process.h"
+#include "brave/browser/ui/webui/brave_webui_source.h"
+#include "brave/components/brave_adblock/adblock_internals/resources/grit/brave_adblock_internals_generated_map.h"
+#include "brave/components/brave_shields/browser/ad_block_service.h"
+#include "components/grit/brave_components_resources.h"
+#include "content/public/browser/web_ui.h"
+#include "content/public/browser/web_ui_controller.h"
+#include "content/public/browser/web_ui_data_source.h"
+#include "content/public/browser/web_ui_message_handler.h"
+#include "services/resource_coordinator/public/cpp/memory_instrumentation/memory_instrumentation.h"
```
Comment (goodov) [inline browser/ui/webui/brave_adblock_internals_ui.cc:25]: pls ensure dependencies you use here are reflected in gn files. `browser/ui` has `check_includes = false`, which makes it pretty scary to add stuff like this without manual deps revision.

File: build/rust/Cargo.lock (hunk 6 above)
Comment (atuchin-m) [reply build/rust/Cargo.lock:7]: yep, see comments

File: browser/ui/webui/brave_adblock_internals_ui.cc (hunk 8 above)
Comment (atuchin-m) [reply browser/ui/webui/brave_adblock_internals_ui.cc:25]: Thanks, done.

File: components/brave_shields/browser/ad_block_service.cc [hunk 9]
```diff
@@ -313,6 +323,42 @@ void AdBlockService::EnableTag(const std::string& tag, bool enabled) {
                      base::Unretained(default_engine_.get()), tag, enabled));
 }
 
+void AdBlockService::GetDebugInfoAsync(GetDebugInfoCallback callback) {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+
+  // base::Unretained() is safe because |default_engine_| is deleted
```
Comment (bridiver) [inline components/brave_shields/browser/ad_block_service.cc:329]: this is not safe because you're passing a raw pointer and when `default_engine_` get deleted this will be a uaf

Comment (atuchin-m) [reply components/brave_shields/browser/ad_block_service.cc:329]: `default_engine_` is always deleted on the same sequence, so it's legal:
https://github.com/brave/brave-core/blob/master/components/brave_shields/content/browser/ad_block_service.cc#L308

Comment (bridiver) [reply components/brave_shields/browser/ad_block_service.cc:329]: nvm, I see what you're saying that this task will necessarily post before the task that deletes default_engine_
